// tailwindcss/lib/cli/commands/init.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.run = run;
exports.optionMap = exports.options = exports.description = exports.usage = void 0;

var constants = _interopRequireWildcard(require("../../constants"));

var colors = _interopRequireWildcard(require("../colors"));

var emoji = _interopRequireWildcard(require("../emoji"));

var utils = _interopRequireWildcard(require("../utils"));

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = Object.defineProperty && Object.getOwnPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : {}; if (desc.get || desc.set) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } } newObj.default = obj; return newObj; } }

const usage = 'init [file]';
exports.usage = usage;
const description = 'Creates Tailwind config file. Default: ' + colors.file(utils.getSimplePath(constants.defaultConfigFile));
exports.description = description;
const options = [{
  usage: '--full',
  description: 'Generate complete configuration file.'
}];
exports.options = options;
const optionMap = {
  full: ['full']
  /**
   * Runs the command.
   *
   * @param {string[]} cliParams
   * @param {object} cliOptions
   * @return {Promise}
   */

};
exports.optionMap = optionMap;

function run(cliParams, cliOptions) {
  return new Promise(resolve => {
    utils.header();
    const full = cliOptions.full;
    const file = cliParams[0] || constants.defaultConfigFile;
    const simplePath = utils.getSimplePath(file);
    utils.exists(file) && utils.die(colors.file(simplePath), 'already exists.');
    const stubFile = full ? constants.defaultConfigStubFile : constants.simpleConfigStubFile;
    utils.copyFile(stubFile, file);
    utils.log();
    utils.log(emoji.yes, 'Created Tailwind config file:', colors.file(simplePath));
    utils.footer();
    resolve();
  });
}
